### ISSUE TITLE
[Proposal] Add a `constrain` function similar to `identity`

### ISSUE THREAD

**Issue by Validark** [roblox-ts/roblox-ts#2080]
Currently, the compiler supports outputting `identity` as a no-op:
```ts
/** Returns the passed in value. The return type matches the type argument.
 * e.g. `identity<1 | 2 | 3>(3)` returns a value of type `1 | 2 | 3`
 */
declare function identity<T>(arg: T): T;
```

However, `identity` changes the type of the argument to match the type parameter. I.e., it loses what the real type is in favor of the type parameter. When one does not want to do that, one has to use a helper like so:
```ts
/** Returns a new identity function with a type argument `T` which returns the passed in literal type which must be assignable to `T`.
 * e.g. `constrain<1 | 2 | 3>()(3)` returns a value of `3`
 */
declare function constrain<T>(): <Q extends T>(arg: Q) => Q;
```

I propose to add `constrain` to [callMacros.d.ts](https://github.com/roblox-ts/compiler-types/blob/master/types/callMacros.d.ts) and make the compiler emit a no-op.

**Comment by Dionysusnu** [comment]
Would it work with just one function call?
```ts
declare function constrain<T, Q extends T = never>(arg: Q): Q;
```

Did some experimenting but couldn't quite get the inference right.

**Comment by Dionysusnu** [comment]
Since 4.9, this is now a TS feature! Instead of a function, it's a dedicated `satisfies` operator.
Example usage
```ts
const a = {
  foo: "bar",
  cake: "lie",
} satisfies Record<string, string>;
type Keys = keyof typeof a; // "foo" | "cake"
```

**Comment by Validark** [comment]
It's about time! This is a good feature for TypeScript. I will leave this issue open until it is ceremoniously closed by #2195
